File: WebTier/src/main/java/com/cc/webtier/helper/sqs/SQSHelper.java
package com.cc.webtier.helper.sqs;

import com.amazonaws.auth.AWSStaticCredentialsProvider;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
import com.amazonaws.services.sqs.model.GetQueueAttributesRequest;
import com.amazonaws.services.sqs.model.Message;
import com.amazonaws.services.sqs.model.ReceiveMessageRequest;
import com.cc.webtier.properties.WebTierProperties;
import io.vavr.control.Try;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

@Component
@Slf4j
public class SQSHelper {
    private final WebTierProperties webTierProperties;

    public SQSHelper(WebTierProperties webTierProperties) {
        this.webTierProperties = webTierProperties;
    }

    public Optional<String> getSQSMessageCount() {
        var attributesRequest = new GetQueueAttributesRequest()
                .withQueueUrl(webTierProperties.getRequestQueue())
                .withAttributeNames("ApproximateNumberOfMessages");

        return Try.of(() -> {
            var amazonSQS = AmazonSQSClientBuilder.standard()
                    .withCredentials(new AWSStaticCredentialsProvider(webTierProperties.getAwsCredentials()))
                    .withRegion(Regions.US_EAST_1)
                    .build();
                    var attributes = amazonSQS.getQueueAttributes(attributesRequest)
                            .getAttributes();
                    log.info("Getting Message Count from Request Queue");
                    amazonSQS.shutdown();
                    return Optional.of(attributes.get("ApproximateNumberOfMessages"));
                }).onFailure(ex -> log.error(ex.getMessage()))
                .get();
    }

    public void publishMessage(String message) {
        Try.run(() -> {
                    var amazonSQS = AmazonSQSClientBuilder.standard()
                            .withCredentials(new AWSStaticCredentialsProvider(webTierProperties.getAwsCredentials()))
                            .withRegion(Regions.US_EAST_1)
                            .build();
                    amazonSQS.sendMessage(webTierProperties.getRequestQueue(), message);
                    log.info("Sent Image details to Queue");
                    amazonSQS.shutdown();
                })
                .onFailure(ex -> log.error(ex.getMessage()));
    }

    public List<Message> readMessagesFromSQS() {
        ReceiveMessageRequest receiveMessageRequest = new ReceiveMessageRequest()
                .withQueueUrl(webTierProperties.getResponseQueue())
                .withMaxNumberOfMessages(10)
                .withWaitTimeSeconds(15);
        var amazonSQS = AmazonSQSClientBuilder.standard()
                .withCredentials(new AWSStaticCredentialsProvider(webTierProperties.getAwsCredentials()))
                .withRegion(Regions.US_EAST_1)
                .build();
        try {
            return amazonSQS.receiveMessage(receiveMessageRequest).getMessages();
        } catch (Exception e) {
            throw new RuntimeException(e);
        } finally {
            amazonSQS.shutdown();
        }
    }

    public void deleteMessage(Message message) {
        Try.run(() -> {
            var amazonSQS = AmazonSQSClientBuilder.standard()
                    .withCredentials(new AWSStaticCredentialsProvider(webTierProperties.getAwsCredentials()))
                    .withRegion(Regions.US_EAST_1)
                    .build();
            amazonSQS.deleteMessage(webTierProperties.getResponseQueue(), message.getReceiptHandle());
            log.info("Deleting Message from the Response Queue");
            amazonSQS.shutdown();
        }).onFailure(ex -> log.error(ex.getMessage()));
    }
}

File: WebTier-Flask/helper/SQSHelper.py
import boto3
from config import Config

properties = Config()


def get_SQS_message_count():
    sqs = boto3.client(
        'sqs',
        aws_access_key_id=properties.get_property("ACCESS_KEY"),
        aws_secret_access_key=properties.get_property("SECRET_KEY"),
        region_name='us-east-1'
    )
    try:
        print("Getting Message Count from Request Queue")
        queue_attributes = sqs.get_queue_attributes(
            QueueUrl=properties.get_property('REQUEST_SQS'),
            AttributeNames=[
                'All'
            ]
        )
        sqs.close()
        return queue_attributes['Attributes']['ApproximateNumberOfMessages']

    except Exception as e:
        print(e)
        return e


def publish_message(message):
    sqs = boto3.client(
        'sqs',
        aws_access_key_id=properties.get_property("ACCESS_KEY"),
        aws_secret_access_key=properties.get_property("SECRET_KEY"),
        region_name='us-east-1'
    )
    try:
        sqs.send_message(
            QueueUrl=properties.get_property('REQUEST_SQS'),
            MessageBody=message,
        )
        print("Sent Image details to Queue: " + message)

    except Exception as e:
        print(e)
        return e

    finally:
        sqs.close()


def read_messages_from_SQS():
    sqs = boto3.client(
        'sqs',
        aws_access_key_id=properties.get_property("ACCESS_KEY"),
        aws_secret_access_key=properties.get_property("SECRET_KEY"),
        region_name='us-east-1'
    )

    try:
        response = sqs.receive_message(
            QueueUrl=properties.get_property('RESPONSE_SQS'),
            MaxNumberOfMessages=10,
            WaitTimeSeconds=15
        )
        return response.get('Messages', [])

    except Exception as e:
        print(e)
        return e

    finally:
        sqs.close()


def delete_message(message):
    sqs = boto3.client(
        'sqs',
        aws_access_key_id=properties.get_property("ACCESS_KEY"),
        aws_secret_access_key=properties.get_property("SECRET_KEY"),
        region_name='us-east-1'
    )

    try:
        sqs.delete_message(
            QueueUrl=properties.get_property('RESPONSE_SQS'),
            ReceiptHandle=message['ReceiptHandle']
        )

    except Exception as e:
        print(e)

    finally:
        sqs.close()
